plonky2/src/hash/merkle_tree.rs

```diff
@@ -145,6 +217,288 @@ fn fill_digests_buf<F: RichField, H: Hasher<F>>(
             subtree_cap.write(fill_subtree::<F, H>(subtree_digests, subtree_leaves));
         },
     );
+
+    // TODO - debug code - to remove in future
+    /*
+    let digests_count: u64 = digests_buf.len().try_into().unwrap();
+    let leaves_count: u64 = leaves.len().try_into().unwrap();
+    let cap_height: u64  = cap_height.try_into().unwrap();
+    let leaf_size: u64 = leaves[0].len().try_into().unwrap();
+    let fname = format!("cpu-{}-{}-{}-{}.txt", digests_count, leaves_count, leaf_size, cap_height);
+    let mut file = File::create(fname).unwrap();
+    for digest in digests_buf {
+        unsafe {
+            let hash = digest.assume_init().to_vec();
+            for x in hash {
+                let str = format!("{} ", x.to_canonical_u64());
+                file.write_all(str.as_bytes());
+            }
+        }
+        file.write_all(b"\n");
+    }
+    */
+}
+
+#[cfg(feature = "cuda")]
+#[repr(C)]
+union U8U64 {
+    f1: [u8; 32],
+    f2: [u64; 4],
+}
+
+#[cfg(feature = "cuda")]
+fn fill_digests_buf_gpu_v1<F: RichField, H: Hasher<F>>(
+    digests_buf: &mut [MaybeUninit<H::Hash>],
+    cap_buf: &mut [MaybeUninit<H::Hash>],
+    leaves: &[Vec<F>],
+    cap_height: usize,
+) {
+    let digests_count: u64 = digests_buf.len().try_into().unwrap();
+    let leaves_count: u64 = leaves.len().try_into().unwrap();
+    let caps_count: u64 = cap_buf.len().try_into().unwrap();
+    let cap_height: u64 = cap_height.try_into().unwrap();
+    let leaf_size: u64 = leaves[0].len().try_into().unwrap();
+    let hash_size: u64 = H::HASH_SIZE.try_into().unwrap();
+
+    let _lock = gpu_lock.lock().unwrap();
+
+    unsafe {
+        let now = Instant::now();
+        fill_init(
+            digests_count,
+            leaves_count,
+            caps_count,
+            leaf_size,
+            hash_size,
+            H::HASHER_TYPE as u64,
+        );
+        print_time(now, "fill init");
+        let now = Instant::now();
+
+        // copy data to C
```

Comment:
we could use rust to copy directly?

Reply:
fill_digests_buf_gpu_v2() copies directly from Rust to the GPU, but the performance is lower compared to Rust -> C -> GPU. We need to investigate the reason

Reply:
Fixed this, now the performance is almost the same (maybe only 1ms difference).